File: MVPJava/app/src/main/java/com/vascome/fogtail/ui/list/ListAppFragment.java
package com.vascome.fogtail.ui.list;

import android.content.Context;
import android.content.Intent;
import android.databinding.DataBindingUtil;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v7.widget.LinearLayoutManager;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.vascome.fogtail.FogtailApplication;
import com.vascome.fogtail.R;
import com.vascome.fogtail.api.entities.RecAreaItem;
import com.vascome.fogtail.databinding.RecyclerRefreshableViewFragmentBinding;
import com.vascome.fogtail.models.AnalyticsModel;
import com.vascome.fogtail.models.AppImageLoader;
import com.vascome.fogtail.models.RecAreaItemsModel;
import com.vascome.fogtail.ui.base.fragments.BaseFragment;
import com.vascome.fogtail.ui.collectionbase.CollectionAreaItemListener;
import com.vascome.fogtail.ui.collectionbase.CollectionPresenter;
import com.vascome.fogtail.ui.collectionbase.ICollectionView;
import com.vascome.fogtail.ui.detail.RecAreaItemDetailActivity;
import com.vascome.fogtail.ui.list.adapter.ListAreaAdapter;
import com.vascome.fogtail.ui.list.adapter.VerticalSpaceItemDecoration;

import java.util.List;

import javax.inject.Inject;

import dagger.Module;
import dagger.Provides;
import dagger.Subcomponent;

/**
 * Created by vasilypopov on 12/5/17
 * Copyright (c) 2017 MVPJava. All rights reserved.
 */

public class ListAppFragment extends BaseFragment implements ICollectionView, CollectionAreaItemListener {

    Context appContext;
    ListAreaAdapter listAreaAdapter;

    @Inject
    CollectionPresenter presenter;

    @Inject
    AppImageLoader networkBitmapClient;

    RecyclerRefreshableViewFragmentBinding binding;

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        appContext = getActivity().getApplicationContext();
        FogtailApplication.get(appContext).applicationComponent().plus(new ListFragmentModule()).inject(this);
    }

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        binding = DataBindingUtil.inflate(inflater, R.layout.recycler_refreshable_view_fragment, container, false);
        return binding.getRoot();
    }


    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        initRecyclerView();
    }

    @Override
    public void onResume() {
        super.onResume();
        presenter.reloadData();
    }

    @Override
    public void setLoadingIndicator(boolean active) {

        if (getView() == null) {
            return;
        }
        binding.recyclerViewSwipeRefresh.post(() -> binding.recyclerViewSwipeRefresh.setRefreshing(active));
    }

    @Override
    public void showItems(List<RecAreaItem> items) {
        runOnUiThreadIfFragmentAlive(() -> {
            setLoadingIndicator(false);
            binding.itemsLoadingErrorUi.setVisibility(View.GONE);
            binding.recyclerViewSwipeRefresh.setVisibility(View.VISIBLE);
            if (listAreaAdapter != null) {
                listAreaAdapter.setData(items);
            }
        });
    }

    @Override
    public void showError() {

        runOnUiThreadIfFragmentAlive(() -> {
            setLoadingIndicator(false);
            binding.recyclerViewSwipeRefresh.setVisibility(View.GONE);
            binding.itemsLoadingErrorUi.setVisibility(View.VISIBLE);
        });

    }

    private void initRecyclerView() {

        binding.recyclerView.addItemDecoration(new VerticalSpaceItemDecoration((int) getResources().getDimension(R.dimen.list_item_vertical_space_between_items)));
        LinearLayoutManager llm = new LinearLayoutManager(appContext, LinearLayoutManager.VERTICAL, false);
        binding.recyclerView.setLayoutManager(llm);
        listAreaAdapter = new ListAreaAdapter(getActivity().getLayoutInflater(), networkBitmapClient, this);
        binding.recyclerView.setAdapter(listAreaAdapter);
        binding.recyclerViewSwipeRefresh.setOnRefreshListener(()->presenter.reloadData());

        presenter.bindView(this);
    }

    @Override
    public void onDestroyView() {
        presenter.unbindView(this);
        binding.unbind();
        super.onDestroyView();
    }

    @Override
    public void onItemClick(RecAreaItem clickedItem) {
        Intent intent = new Intent(appContext, RecAreaItemDetailActivity.class);
        intent.putExtra("item", clickedItem);
        startActivity(intent);
    }

    @Subcomponent(modules = ListFragmentModule.class)
    public interface ListFragmentComponent {
        void inject(@NonNull ListAppFragment itemsFragment);
    }

    @Module
    public class ListFragmentModule {

        @Provides
        @NonNull
        public CollectionPresenter provideItemsPresenter(@NonNull RecAreaItemsModel itemsModel,
                                                         @NonNull AnalyticsModel analyticsModel) {
            return new CollectionPresenter(itemsModel, analyticsModel);
        }
    }

}

File: MosbyMVIKotlin/app/src/main/java/com/vascome/fogtail/presentation/main/fragment/list/ListAppFragment.kt
package com.vascome.fogtail.presentation.main.fragment.list

import android.os.Bundle
import android.support.v7.widget.LinearLayoutManager
import android.view.LayoutInflater
import android.view.View
import android.view.View.GONE
import android.view.View.VISIBLE
import android.view.ViewGroup
import android.view.animation.AnimationUtils
import com.jakewharton.rxbinding2.support.v4.widget.RxSwipeRefreshLayout
import com.jakewharton.rxbinding2.view.clicks

import com.vascome.fogtail.R
import com.vascome.fogtail.data.network.AppImageLoader
import com.vascome.fogtail.presentation.base.fragments.BaseFragment
import com.vascome.fogtail.presentation.detail.RecAreaItemDetailActivity
import com.vascome.fogtail.presentation.main.CollectionPresenter
import com.vascome.fogtail.presentation.main.CollectionView
import com.vascome.fogtail.presentation.main.CollectionViewState
import com.vascome.fogtail.presentation.main.domain.model.RecAreaItem
import com.vascome.fogtail.presentation.main.fragment.list.adapter.ListAreaAdapter
import com.vascome.fogtail.presentation.main.fragment.list.adapter.VerticalSpaceItemDecoration
import com.vascome.fogtail.presentation.main.utils.CollectionAreaItemListener
import io.reactivex.Observable
import kotlinx.android.synthetic.main.recycler_refreshable_view_fragment.*
import timber.log.Timber
import java.util.concurrent.TimeUnit

import javax.inject.Inject

@Suppress("MemberVisibilityCanPrivate")
/**
 * Created by vasilypopov on 12/5/17
 * Copyright (c) 2017 MVPJava. All rights reserved.
 */

class ListAppFragment :
        BaseFragment<CollectionView, CollectionPresenter>(),
        CollectionView,
        CollectionAreaItemListener {
    @Inject
    lateinit var collectionPresenter: CollectionPresenter

    @Inject
    lateinit var imageLoader: AppImageLoader

    private val listAreaAdapter by lazy {
        ListAreaAdapter(activity.layoutInflater, imageLoader, this)
    }

    override fun createPresenter(): CollectionPresenter = collectionPresenter

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.recycler_refreshable_view_fragment, container, false)
    }


    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        initRecyclerView()
    }

    private fun initRecyclerView() {

        recyclerView.addItemDecoration(VerticalSpaceItemDecoration(resources.getDimension(R.dimen.list_item_vertical_space_between_items).toInt()))
        val llm = LinearLayoutManager(context, LinearLayoutManager.VERTICAL, false)
        recyclerView.layoutManager = llm
        recyclerView.adapter = listAreaAdapter
    }

    override fun loadOnStartIntent(): Observable<Boolean> {
        return Observable.just(true).doOnComplete { Timber.d("start loading completed") }
    }

    override fun pullToRefreshIntent(): Observable<Boolean> {
        return RxSwipeRefreshLayout.refreshes(swipeRefreshLayout).map({ _ -> true })
    }

    override fun retryButtonClickIntent(): Observable<Boolean> {
        return tryAgainButton.clicks()
                .debounce(500, TimeUnit.MICROSECONDS)
                .map { _ -> true }
    }

    override fun render(viewState: CollectionViewState) {
        Timber.d("render %s", viewState)

        when {
            viewState.loading -> renderLoading()
            viewState.data != null -> renderData(viewState)
            viewState.error != null -> renderError()
        }
    }

    private fun renderError() {
        swipeRefreshLayout.post { swipeRefreshLayout.isRefreshing = false }
        swipeRefreshLayout.visibility = GONE
        errorContainer.visibility = VISIBLE
        listAreaAdapter.setData(emptyList())
    }

    private fun renderData(viewState: CollectionViewState) {
        swipeRefreshLayout.post { swipeRefreshLayout.isRefreshing = false }
        errorContainer.visibility = GONE
        swipeRefreshLayout.visibility = VISIBLE
        listAreaAdapter.setData(viewState.data!!)
        if(!isRestoringViewState) {
            runLayoutAnimation()
        }
    }

    private fun renderLoading() {
        errorContainer.visibility = GONE
        swipeRefreshLayout.visibility = VISIBLE
        swipeRefreshLayout.post { swipeRefreshLayout.isRefreshing = true }
    }

    override fun onItemClick(clickedItem: RecAreaItem) {
        RecAreaItemDetailActivity.start(activity, clickedItem)
    }

    private fun runLayoutAnimation() {
        AnimationUtils.loadLayoutAnimation(recyclerView.context, R.anim.layout_animation_fall_down).apply {
            recyclerView.layoutAnimation = this
        }
        recyclerView.scheduleLayoutAnimation()
    }
}
